factor/sms/classes/factor.php

```diff
@@ -0,0 +1,358 @@
+        }
+
+        return $return;
+    }
+
+    /**
+     * SMS Factor implementation.
+     *
+     * {@inheritDoc}
+     */
+    public function setup_factor_form_definition($mform) {
+        global $USER;
+        if (empty($USER->phone2)) {
+            $mform->addElement('hidden', 'verificationcode', 0);
+            $mform->setType("verificationcode", PARAM_ALPHANUM);
+
+            // Add field for phone number setup.
+            $mform->addElement('text', 'phonenumber', get_string('addnumber', 'factor_sms'));
+            $mform->setType('phonenumber', PARAM_TEXT);
+        }
+    }
+
+    /**
+     * SMS Factor implementation.
+     *
+     * {@inheritDoc}
+     */
+    public function setup_factor_form_definition_after_data($mform) {
+        global $USER, $SESSION;
+
+        // Nothing if they dont have a number added.
+        if (empty($USER->phone2)) {
+            return $mform;
+        }
+
+        $mform->addElement(new \tool_mfa\local\form\verification_field());
+        $mform->setType('verificationcode', PARAM_ALPHANUM);
+
+        if (empty($SESSION->mfa_sms_setup_code)) {
+            // They need a code to verify this number.
+            // Lets generate a code and send it in place, we don't need this to get to the DB.
+            $code = random_int(100000, 999999);
+
+            $this->sms_verification_code($code, $USER->phone2);
+            // Store as string for form comparison.
+            $SESSION->mfa_sms_setup_code = (string) $code;
+            $mform = $this->add_redacted_sent_message($mform);
+        } else {
+            // Dont generate or send, just tell them you did!
+            $mform = $this->add_redacted_sent_message($mform);
+        }
+    }
+
+    /**
+     * SMS Factor implementation.
+     *
+     * {@inheritDoc}
+     */
+    public function setup_factor_form_validation($data) {
+        global $SESSION, $USER;
+
+        // No validation on raw number.
+        if (empty($USER->phone2)) {
+            return [];
+        }
+
+        $errors = [];
+        if ($data['verificationcode'] !== $SESSION->mfa_sms_setup_code) {
+            $errors['verificationcode'] = get_string('wrongcode', 'factor_sms');
+        }
+
+        return $errors;
+    }
+
+    /**
+     * Adds an instance of the factor for a user, from form data.
+     *
+     * {@inheritDoc}
+     */
+    public function setup_user_factor($data) {
+        global $DB, $SESSION, $USER;
+
+        // Handle phone number submission.
+        if (empty($USER->phone2)) {
+            $DB->set_field('user', 'phone2', $data->phonenumber, ['id' => $USER->id]);
+            // Also update $USER.
+            $USER->phone2 = $data->phonenumber;
+            \core\notification::info(get_string('numberadded', 'factor_sms'));
+            $addurl = new \moodle_url('/admin/tool/mfa/action.php', [
+                'action' => 'setup',
+                'factor' => 'sms',
+                'sesskey' => sesskey(),
```

Comment:
remove sesskey? I don't think we need it here and we don't want it in the url after?

Reply:
Sesskey is required here, the page clears the URL once loaded, so it never really appears in the url